src/lib/y2partitioner/widgets/apqn_selector.rb

```diff
@@ -0,0 +1,83 @@
+# Copyright (c) [2020] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "yast2/popup"
+require "cwm"
+require "y2partitioner/actions/controllers/encryption"
+
+module Y2Partitioner
+  module Widgets
+    # Widget to select APQNs for generating a new secure key for pervasive encryption
+    class ApqnSelector < CWM::MultiSelectionBox
+      # Constructor
+      #
+      # @param controller [Actions::Controllers::Encryption]
+      # @param enable [Boolean] whether the widget should be enabled on init
+      def initialize(controller, enable: true)
+        textdomain "storage"
+
+        @controller = controller
+        @enable_on_init = enable
+      end
+
+      # @return [String]
+      def label
+        _("Available APQNs:")
+      end
+
+      # @macro seeAbstractWidget
+      def init
+        enable_on_init ? enable : disable
+        self.value = controller.apqns
+      end
+
+      # @return [Array<String, String>]
+      def items
+        controller.online_apqns.map { |d| [d.name, d.name] }
```

Comment:
I see this is a common pattern, so I guess it is OK as it is. However, I do not see the point in having to rely on a controller to get the items. Why do not pass the list of apqns in the constructor? IMHO, it would make the widget more reusable: it simply gets a list of apqns and the `#value` method returns the selected ones. No controller needed at all except for `#store`, and that could takes us to another discussion :)

Reply:
I don't like either to have a widget depending on a controller. But as you say, it is the common pattern we have decided to follow to avoid having logic scattered in all widgets around. Anyway, this widget will not be re-usable without a controller, so I would prefer to simply pass the controller in the constructor (as for the rest of widgets) and use that controller to get/set all the information.